### PR TITLE
[SPARK-40713][CONNECT] Improve SET operation support in the proto and the server

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

1. Rename `Union` to `SetOperation` and make it cover `except`, `intersect`, `union`.
2. Improve server side support for `SetOperation`.

### Why are the changes needed?

Improve SET operation support in the proto and the server.

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

UT

### CODE REVIEW

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 1]
```diff
@@ -107,15 +107,17 @@ message Join {
   }
 }
 
-// Relation of type [[Union]], at least one input must be set.
```
Comment (amaliujia) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:110]: This documentation (`at least one input must be set`) does not match the implementation `assert(u.getInputsCount == 2, "Union must have 2 inputs")`. To avoid confusion I choose to remove this for now.

Once our implementation is stable, we can go back to re-document the semantic.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 2]
```diff
@@ -35,7 +35,7 @@ message Relation {
     Project project = 3;
     Filter filter = 4;
     Join join = 5;
-    Union union = 6;
```
Comment (grundprinzip) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:38]: Why is Set better than Union? Just from a SQL naming persepective, Union might be what users are used to?

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 1 above)
Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:110]: TBH i think the comment was not correct in the first place. I mean while a set can have one element, the question is does it make sense to have a set plan with only one?

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 3]
```diff
@@ -107,15 +107,17 @@ message Join {
   }
 }
 
-// Relation of type [[Union]], at least one input must be set.
-message Union {
+// Relation of type [[SetOperation]]
+message SetOperation {
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:111]: can we have 3 individual plans?

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 2 above)
Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:38]: I am pretty sure`Set operation` is the SQL word to describe `union` `intersect` `except`. But Wenchen suggests to have individual plans for each, in that it won't matter.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 3 above)
Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:111]: Of course union supports `ALL`. Not sure why we would need three plans for this?

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:111]: Because catalyst has 3 plans for set operations, I think it's better to follow it in proto so that they can evolve separately.

To @grundprinzip : we do support `UNION ALL`, but the logical plan for `UNION` is `Distinct(Union(...))`. The `Union` operator does not have a `isAll` flag.

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:111]: After a second thought, if we treat the proto layer as a standard for general SQL query plans, it doesn't have to be consistent with catalyst. Then a single `SetOperation` LGTM.

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:111]: Note: Spark has a tricky optimization for a long chain of `Union`. e.g. `df1.union(df2).union(df3).union...`, Spark does not create a left-deep `Union` tree (which makes analyzer/optimizer run very slow), but just a single `Union` with many children. I'm not sure if we want to do this optimization in Spark Connect clients or not. But if we do, we need a special `Union` proto definition which takes a Seq of children, not just left and right.

Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:111]: Ok maybe we keep the same message as `SetOperation` for now. When we face the need for further optimization on specific case, we can propose ways to help that.

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:111]: nvm, Spark Connect DataFrame operations combine parsed plan, not analyzed plan, so we don't need this early Union optimization. A single `SetOperation` should work.

Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:111]: That would be great then!

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 1 above)
Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:111]: I am thinking it should be always two inputs so the implementation should be correct. Though I don't know enough about dataframe API usage so not sure if there could be only one, or more than two, etc.

Given proto is also a API layer, before 3.4 release I think we need to add a lot more comments on proto to clarify each one's semantic. We can finalize how many inputs are possible at that time.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 4]
```diff
@@ -108,15 +108,17 @@ message Join {
...
   repeated Relation inputs = 1;
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:113]: shall we require 2 inputs `left` and `right`? Our SQL parser only generates set operation with 2 children.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 5]
```diff
@@ -108,15 +108,17 @@ message Join {
...
-    UNION_TYPE_UNSPECIFIED = 0;
-    UNION_TYPE_DISTINCT = 1;
-    UNION_TYPE_ALL = 2;
+  SetOpType set_op_type = 2;
+  bool is_all = 3;
+
+  enum SetOpType {
+    SET_OP_TYPE_UNSPECIFIED = 0;
+    SET_OP_TYPE_INTERSECT = 1;
+    SET_OP_TYPE_ALL = 2;
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:120]:
```suggestion
    SET_OP_TYPE_UNION = 2;
```

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 6]
```diff
@@ -54,9 +55,9 @@ package object dsl {
       }
 
       def join(
-          otherPlan: proto.Relation,
-          joinType: JoinType = JoinType.JOIN_TYPE_INNER,
-          condition: Option[proto.Expression] = None): proto.Relation = {
+        otherPlan: proto.Relation,
```
Comment (cloud-fan) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:58]: nit: 4 spaces indentation for parameter list.

Comment (amaliujia) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:58]: thanks.

My IDE always breaks it.... let me figure out a way to configure IDE to avoid this happening again.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 4 above)
Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:113]: I see
```
queryTerm
    : queryPrimary                                                                       #queryTermDefault
    | left=queryTerm {legacy_setops_precedence_enabled}?
        operator=(INTERSECT | UNION | EXCEPT | SETMINUS) setQuantifier? right=queryTerm  #setOperation
    | left=queryTerm {!legacy_setops_precedence_enabled}?
        operator=INTERSECT setQuantifier? right=queryTerm                                #setOperation
    | left=queryTerm {!legacy_setops_precedence_enabled}?
        operator=(UNION | EXCEPT | SETMINUS) setQuantifier? right=queryTerm              #setOperation
```

Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto]: Done. Now the definition is easier to follow: just check if there is left or right.

**Review by cloud-fan — APPROVED** [review]

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala [hunk 7]
```diff
@@ -81,6 +81,31 @@ class SparkConnectProtoSuite extends PlanTest with SparkConnectPlanTest {
...
+        import org.apache.spark.sql.connect.dsl.plans._
+        transform(connectTestRelation.intersect(connectTestRelation, isAll))
+      }
+      val sparkPlan2 = sparkTestRelation.intersect(sparkTestRelation, isAll)
+      comparePlans(connectPlan2.analyze, sparkPlan2.analyze, false)
+    }
+
+    val connectPlan3 = {
+      import org.apache.spark.sql.connect.dsl.plans._
+      transform(connectTestRelation.union(connectTestRelation))
```
Comment (cloud-fan) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:103]: last question: does the current spark connect client API support specifying `isAll` for Union?

Comment (amaliujia) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:103]: My understanding is no:

there are 3 unions: `unionAll`, `unionByName` and `union distinct`. For the third case DataFrame API says to use `DISTINCT(UNION(....))`. For the `UNIONAll` itself there is no flag for `isAll` so by default `Union` is `UnionAll`

There should be some context for this difference but I cannot tell why now.

Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:103]: This is one of the cases where we shouldn't just blindly take what the DataFrame API does but we should think in terms of generalized parse plans. 


In this case I much rather have the option on the union for a All rather than doing something implicit.

Comment (amaliujia) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:103]: We are doing Union with isAll because along with intersect and except in the same message.

I don't agree with it though. Generally speaking if there is something DataFrame decides to do, I think figuring out the context will be useful. People working on Spark are smart. They can of course just does union with isAll but they decided to not do so. There must be a valid reason and that reason might also apply to connect.

Comment (amaliujia) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:113]: Another reason actually to follow Catalyst: our clients, by best effort, should follow existing data frame API to achieve changing imports and rest of code work. 

So it's expected that users will use union().distinct() to construct the case above. If we ask clients to smartly identify these two API can be coalesce together, that will be too much for clients. Instead clients just convert each API to a plan node and connect proto follow Catalyst then it will work.

Comment (cloud-fan) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:113]: Spark Connect may have many clients, built-in or third-party. I think it's better to make the query plan proto definition general/standard.

It was kind of a mistake that we forgot to add DataFrame APIs for users to specify `ALL | DISTINCT` for set operations. We added the `isAll` flag to the plan later, because the SQL parser started to support it.

`byName` flag is only supported by Union now, but I think other two set operators should support it as well.

To summarize by proposal:
1. add a `byName` flag to `SetOperation` proto definition
2. The server translate `Union(..., isAll = false)` to `Distinct(Union)`
3. The server fails if `byName = true` and the set type is intercept or except.

Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:113]: Generally, the proto definitions should follow as much as possible the basics for the relational algebra / SQL language. As @cloud-fan mentioned, since there maybe other clients. It's important to not just follow the quirks of the DF API because its the way it's currently implemented.

Having a proper `SetOperation` definition is not stopping us from continuing to have customers call `df.union().distinct()`.

Comment (amaliujia) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:113]: I am fine on the proposal as long as it does not prevent the clients being compatible with existing data frame API. Any extra functionality is fine.

Thanks @cloud-fan for you summary. Let me change this PR.

**Comment by cloud-fan** [comment]
please fix conflicts

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 8]
```diff
@@ -127,15 +127,19 @@ message Join {
...
-    UNION_TYPE_UNSPECIFIED = 0;
-    UNION_TYPE_DISTINCT = 1;
-    UNION_TYPE_ALL = 2;
+// Relation of type [[SetOperation]]
+message SetOperation {
+  Relation left_input = 1;
+  Relation right_input = 2;
+  SetOpType set_op_type = 3;
+  bool is_all = 4;
+  bool union_by_name = 5;
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:136]: just `by_name`, as it applies to all set operations in the future.

**Review by cloud-fan — APPROVED** [review]

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 8 above)
Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto]: makes sense. Done.

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala (hunk 7 above)
Comment (HyukjinKwon) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala]: Hm, sorry for joining this conversation late. Why don't we just match it to SQL plan implementation? If the design is the issue, we can refactor them together?

Comment (cloud-fan) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala]: The issue is, we can't change proto message freely. It's API, while catalyst plans are internal implementations. That's why we need more considerations when adding proto messages, to make them future-proof.
